### ISSUE TITLE
Activation hooks are incorrectly flagged as having invalid names

### ISSUE THREAD

**Issue by samwilson** [WordPress-Coding-Standards/WordPress-Coding-Standards#881]
I'm getting a phpcs error of type `WordPress.NamingConventions.ValidHookName.UseUnderscores` for activation hooks which are created in `register_activation_hook()`. These hooks include dots and slashes because they include filenames; the in-code docs say:

> In the name of this hook, PLUGINNAME is replaced with the name of the plugin, including the optional subdirectory. For example, when the plugin is located in wp-content/plugins/sampleplugin/sample.php, then the name of this hook will become 'activate_sampleplugin/sample.php'.

I think it's possible to add more allowed characters to the above sniff, but that doesn't seem like the optimal solution because then it would also fail to catch other hook names that (correctly) shouldn't include slashes or dots.

**Comment by jrfnl** [comment]
@samwilson Normally you would use something like `add_action( 'activate_' . __FILE__, 'myplugin_activation_routine' );`
Hard-coding the file-name is considered bad practice as an end-user might install the plugin in a folder with a different name than the default and your plugin will break in that case.

Pro-tips:
1 - DON'T use an activation routine at all. Use an upgrade routine instead. This will improve your plugin's compatibility with WP Multi-site and make the it more reliable all round.
For some code examples see: https://speakerdeck.com/jrf/the-secret-to-mastering-wordpress-multi-site-php-world#26 and https://speakerdeck.com/jrf/the-secret-to-mastering-wordpress-multi-site-php-world#27
.
.
.
.
.
99 - If you really, really insist on using an activation routine, follow a pattern along the lines of the below (of course using classes etc, but you get my drift):
```php
function myplugin_activation_routine() {
	// Do things.

	// Allow for hooking into your activation routine without issues.
	do_action( 'myplugin_activiation' );
}
register_activation_hook( __FILE__, 'myplugin_activation_routine' );
```

**Comment by samwilson** [comment]
That makes sense @jrfnl, especially with regard to being more compatible with multi-site. But I just wonder about marking as a coding standard violation something that is explicitly mentioned in the [Plugin Handbook](https://developer.wordpress.org/plugins/the-basics/activation-deactivation-hooks/#activation). There seems to be a conflict there.

Also, one thing I like about activation hooks is that, rather than relying on a simple version check, it's possible to actually do in-depth (i.e. sometimes expensive) checks of the plugin's set-up, and upgrade accordingly. For example, rather than assuming a new table column was created in v.2.1 of the plugin, actually checking for its existence — not something that one would want to do on every page load.

**Comment by jrfnl** [comment]
@samwilson I can't find any reference to the resulting hookname in the Plugin Handbook and all the examples use the `__FILE__` constant just like I mentioned before, so I don't see the conflict.

The function documentation for the `register_activation_hook()` does mention it, but that is another matter.

> one thing I like about activation hooks is that, rather than relying on a simple version check, it's possible to actually do in-depth (i.e. sometimes expensive) checks of the plugin's set-up, and upgrade accordingly. 

An upgrade routine will generally run a lot more often than an activation routine, so will be a lot more reliable - quite apart from the fact that it will also be run on child sites when a plugin is network activated.

For expensive checks, you could consider periodically running a cron job to make sure things are as they should be. Scheduling a cron job is not expensive and can be done from the upgrade routine, just make sure you schedule it correctly (and don't spawn thousands of them).

**Comment by samwilson** [comment]
No, good point; all I meant was that the handbook tells people to use `register_activation_hook()`.

The error I'm getting is because of the way `register_activation_hook()` constructs the hook name:

```php
function register_activation_hook($file, $function) {
	$file = plugin_basename($file);
	add_action('activate_' . $file, $function);
}
```

So if it's called in `wp-content/plugins/sampleplugin/sample.php` with `register_activation_hook( __FILE__, 'myplugin_activation_routine' );` then the resulting hook name will be `activate_sampleplugin/sample.php` — which according to the above sniff is not valid. That's the crux of this issue.

**Comment by jrfnl** [comment]
@samwilson So on what code does the sniff trigger ? Something like the below ?

`add_action('activate_sampleplugin/sample.php', 'my_function');`

I'm starting to repeat myself, but that is **_bad practice_** as it will break if a user installs in a custom folder.

If you want to hook into that hook, you should do it like this from the same file as where the registration hook is...

`add_action('activate_' . __FILE__, 'my_function');`

...  or add a constant which remembers the value of `__FILE__` when in the main file and use that constant in other files.

```php
<?php
// Main file
define( 'MYPLUGIN_MAIN_FILE', __FILE__ );
register_activation_hook( __FILE__, 'myplugin_activation_routine' );

// Secondary file
<?php
add_action('activate_' . MYPLUGIN_MAIN_FILE, 'my_function');
```

... or use the pattern with a custom hook as shown in my [previous comment](https://github.com/WordPress-Coding-Standards/WordPress-Coding-Standards/issues/881#issuecomment-288152825)

**Comment by jrfnl** [comment]
Anyways, the short of it is, that you should never use a hard-coded hook name like `activate_sampleplugin/sample.php` in your plugin anyway.

The sniff should trigger on it, though maybe with another error message. No matter what, that code should be banned.

**Comment by samwilson** [comment]
> @samwilson So on what code does the sniff trigger ? Something like the below ?

No, my point is that if someone follows the handbook and uses `register_activation_hook( __FILE__, 'myplugin_activation_routine' );` then the code will run foul of this sniff!

Of course it's quite possible to get around this by just using `add_action('activate_myplugin', 'my_function')` and *not* using `register_activation_hook()`, but surely it's better to follow the official handbook?

>Anyways, the short of it is, that you should never use a hard-coded hook name like activate_sampleplugin/sample.php in your plugin anyway.

Just to reiterate, my point is that *I'm not using such a construction in my code*.

Sorry to go on about  this. :-) At the moment, my fix is just to exclude this sniff from running on my main plugin file.

**Comment by jrfnl** [comment]
@samwilson 

> then the code will run foul of this sniff!

No it will not as WP core does not use the hard coded version either. As long as the hard coded version of that particular hook is not encountered, this sniff will not throw errors.

> At the moment, my fix is just to exclude this sniff from running on my main plugin file.

Why ?

Provide me with some legitimate example code on which the sniff triggers - as in a proper code example with real code which doesn't break all the best practices already discussed above, a code sample which could be added to the unit tests, and then I might have another look.

**Comment by samwilson** [comment]
Ah, okay, so this makes more sense now. You're saying that the problem isn't with the *name* of the hook having slashes or dots, but just that it shouldn't be *called* by hard-coded name?

I'm calling the activation hook in my plugin from a test. I'll update it and stop using `register_activation_hook()`.

Thank you for your patience with me!

**Comment by jrfnl** [comment]
> You're saying that the problem isn't with the name of the hook having slashes or dots, but just that it shouldn't be called by hard-coded name?

No, the name is not in accordance with the WP's own coding standard as found in the handbook.
However, for historical reasons I suspect this hook name will not be changed.

All the same, for this particular "wrong" name, if used in the advised way - i.e. using `__FILE__`, not hard-coded - the sniff will not pick up on it, so it shouldn't cause any problems.

**Comment by metamn** [comment]
I have the same warning for `apply_filters`: 

```
add_theme_support( 'custom-background',
		apply_filters( 'more_wp_custom_background_args',
			array(
				'default-color' => 'ffffff',
				'default-image' => '',
			)
		)
	);
```
The message is:

```
[WordPress.NamingConventions.ValidHookName.UseUnderscores] Words in hook names should be separated using underscores. Expected: '_', but found: 'more_wp_custom_background_args'.
```

**Comment by jrfnl** [comment]
@metamn This issue has been closed for over a year and your issue, while related to the same sniff, is not the same issue. Could you open a new issue for this ?

**Comment by jrfnl** [comment]
@metamn Also, when opening the issue, could you a little more info as based on the code snippet you showed, I cannot reproduce the issue.
